control-plane-operator/controllers/hostedcontrolplane/mcs/reconcile.go

```diff
@@ -0,0 +1,143 @@
+package mcs
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/hypershift/control-plane-operator/api"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/pki"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/util"
+	mcfgv1 "github.com/openshift/hypershift/thirdparty/machineconfigoperator/pkg/apis/machineconfiguration.openshift.io/v1"
+)
+
+func ReconcileMachineConfigServerConfig(cm *corev1.ConfigMap, p *MCSParams) error {
+	p.OwnerRef.ApplyTo(cm)
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	serializedDNS, err := serialize(p.DNS)
+	if err != nil {
+		return err
+	}
+	serializedInfra, err := serialize(p.Infrastructure)
+	if err != nil {
+		return err
+	}
+	serializedNetwork, err := serialize(p.Network)
+	if err != nil {
+		return err
+	}
+	serializedProxy, err := serialize(p.Proxy)
+	if err != nil {
+		return err
+	}
+	serializedPullSecret, err := serialize(p.PullSecret)
+	if err != nil {
+		return err
+	}
+	serializedMasterConfigPool, err := serializeConfigPool(masterConfigPool())
+	if err != nil {
+		return err
+	}
+	serializedWorkerConfigPool, err := serializeConfigPool(workerConfigPool())
+	if err != nil {
+		return err
+	}
+
+	cm.Data["root-ca.crt"] = string(p.RootCA.Data[pki.CASignerCertMapKey])
+	cm.Data["combined-ca.crt"] = string(p.CombinedCA.Data[pki.CASignerCertMapKey])
+	cm.Data["cluster-dns-02-config.yaml"] = serializedDNS
+	cm.Data["cluster-infrastructure-02-config.yaml"] = serializedInfra
+	cm.Data["cluster-network-02-config.yaml"] = serializedNetwork
+	cm.Data["cluster-proxy-01-config.yaml"] = serializedProxy
+	cm.Data["install-config.yaml"] = p.InstallConfig.String()
+	cm.Data["pull-secret.yaml"] = serializedPullSecret
+	cm.Data["master.machineconfigpool.yaml"] = serializedMasterConfigPool
+	cm.Data["worker.machineconfigpool.yaml"] = serializedWorkerConfigPool
+	return nil
+}
+
+func serialize(obj client.Object) (string, error) {
+	return util.SerializeResource(obj, api.Scheme)
+}
+
+func serializeConfigPool(obj client.Object) (string, error) {
+	return util.SerializeResource(obj, &machineConfigTyper{})
+}
+
+// NOTE: this is here because we don't import the the machineconfig types into
+// our scheme, and apart from this use case, we don't need to.
+type machineConfigTyper struct{}
+
+func (t *machineConfigTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind, bool, error) {
+	return []schema.GroupVersionKind{
+		{
+			Group:   "machineconfiguration.openshift.io",
+			Version: "v1",
+			Kind:    "MachineConfigPool",
+		},
+	}, false, nil
+}
+
+func (t *machineConfigTyper) Recognizes(gvk schema.GroupVersionKind) bool {
```

Comment:
Erm that seems very liberal. Does the `Recognizes` only get called with kinds under the configured Group?

Reply:
I'm only passing a machineConfig to the serializer. However, I see that this could be handled better by just creating a new scheme, adding the MachineConfig group to it and then using that.